src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -616,6 +619,152 @@ int D4NFilterBucket::list(const DoutPrefixProvider* dpp, ListParams& params, int
+        .cacheObj = object,
+        .blockID = 0,
+        .size = 0,
+      };
+      
+      blocks.push_back(block);
+      objects.push_back(object);
+
+      std::string oid_version;
+      if (c_obj->have_instance()) {
+        oid_version = c_obj->get_instance();
+      } else {
+        oid_version = "null";
+      }
+      off_t lst;
+      std::string head_oid_in_cache;
+      block.cacheObj.objName = "_:" + oid_version + "_" + block.cacheObj.objName;
+      auto it = std::find_if(dir_blocks.begin(), dir_blocks.end(), [&dpp, &oid_version, &block] (const rgw::d4n::CacheBlock& dir_block) { 
+        return (dir_block.cacheObj.objName == (block.cacheObj.objName));
+      });
+      if (it != dir_blocks.end()) {
+        lst = it->cacheObj.size;
+        block.cacheObj.dirty = it->cacheObj.dirty;
+	    head_oid_in_cache = get_cache_block_prefix(c_obj.get(), it->version);
+
+        if (it->cacheObj.dirty) {
+          if (!this->filter->get_policy_driver()->get_cache_policy()->invalidate_dirty_object(dpp, get_cache_block_prefix(c_obj.get(), it->version))) {
+            ldpp_dout(dpp, 10) << "D4NFilterBucket::" << __func__ << "(): Failed to invalidate obj=" << c_obj->get_name() << " in cache" << dendl;
+                return -EINVAL;
+          }
+        }
+      } else {
+        ldpp_dout(dpp, 20) << "D4NFilterBucket::" << __func__ << "(): Listing retrieved from backend for object " << c_obj->get_name() << dendl;
+
+        if (blockDir->get(dpp, &block, y) < 0) {
+          ldpp_dout(dpp, 20) << "D4NFilterBucket::" << __func__ << "(): Failed to retrieve block size, ret=" << ret << dendl;
+          return ret;
+        } // TODO: Alternate to getting size?
+        lst = block.cacheObj.size;
+      }
+
+      // Handle versioned head objects
+      ldpp_dout(dpp, 20) << "D4NFilterBucket::" << __func__ << "(): versioned oid: " << block.cacheObj.objName << dendl;
+      blocks.push_back(block);
+
+      // Handle data blocks
+      ldpp_dout(dpp, 20) << "D4NFilterBucket::" << __func__ << "(): Object size=" << lst << dendl;
+      off_t fst = 0;
+      do {
+        ldpp_dout(dpp, 20) << "D4NFilterBucket::" << __func__ << "(): handling object=" << obj.key << dendl;
+        rgw::d4n::CacheBlock data_block;
+        if (fst >= lst) {
+          break;
+        }
+        off_t cur_size = std::min<off_t>(fst + dpp->get_cct()->_conf->rgw_max_chunk_size, lst);
+        off_t cur_len = cur_size - fst;
+        data_block.cacheObj.bucketName = this->get_bucket_id();
+        data_block.cacheObj.objName = c_obj->get_oid();
+        ldpp_dout(dpp, 20) << "D4NFilterBucket::" << __func__ << "(): data_block=" << data_block.cacheObj.objName << dendl;
+        data_block.size = cur_len;
+        data_block.blockID = fst;
+
+        fst += cur_len;
+        blocks.push_back(data_block); // TODO: How to account for number of data blocks per object in PIPELINE_MAX?
+      } while (fst < lst);
+    }
```

Comment:
can you mark where the do { for iterating through the list results ends? I will continue reviewing once that is marked.